### PR TITLE
[Draft] Circuitrework

### PR DESCRIPTION
**Pull request by MNarath1**
## About The Pull Request

This Draft intends to rework circuits heavily as well as change the current circuit UI to TGUI.
Main Points are of course TGUI, making printer into a machine instead of an item and more akine to a protolath, gives different circuits actual material usage beyond iron, removes upgrade disk and instead uses normal reseach sync(possible multiple different reseach paths too instead of just a single upgrade one, removes insta cloning and makes circuits clone time effected by the printer parts, ore silo link, readds old butched circuits after reworking/rebalancing them, rebalances annyoning spam circuit(like radio circuit etc), fixes some known circuit exploids and adds new circuit saving system instead of the current import cloning that allows to import strings to the server now that data stays saved on the server currently in form of sav files this might change trough. Additional admin tools to better track offenders using circuits etc.

More details will follow but here are some of the main bullet points i wana tackle for now.

- [x] Change current circuit import system to one based on sav data saved on the server(might change to sql)
- [ ] Write Backend for TGUI(still in work)
- [ ] Actually integrating the new TGUI
- [ ] Changing circuit printer to machine(done partially)
- [ ] Reworks material usage for circuits(no longer only iron)
- [ ] reworks the upgrade disks into actual reseach sync with the reseach server instead
- [ ] readds chem circuits and rebalance them so most of them can only be used in bigger assemblies
- [ ] removes insta cloning and reworks cloning times
- [ ] readds some other long butched circuits after extensive rework(possible but don't count on that)
- [ ] allows circuit printer to be linked to ore silo
- [ ] rebalance some of the more problematic/annonying circuits(radio etc)
- [ ] fixes known circuit bugs(aka pins stay connected after circuit has been ejected etc.)

:cl:
Nothing here yet
/:cl:

### CODE REVIEW

**Comment by PowerfulBacon** [comment]
I really dont think unnecessary server side saving is a good idea, it adds more problems by having the files on the server and solves [???]

**Comment by ItsTisme** [comment]
> I really dont think unnecessary server side saving is a good idea, it adds more problems by having the files on the server and solves [???]

Idea was that griefer could not import a grief script instantly on burner accounts IRC

**Comment by PowerfulBacon** [comment]
Griefers can now write text to the server

**Comment by MNarath1** [comment]
> Griefers can now write text to the server

I am going to add some sanitisation as well as the current text filter ontop of it also wanna actuall change this to sql so its more akine to library etc would make managing those circuits easiler too if its actually managed trough an sql.

**Comment by aidan-bowman** [comment]
aw yis i was gonna do exactly this (well some of it). seems you beat me to the draw

actually i have a couple ideas that i think would be very effective that you missed here. wanna team up or something?

**Comment by MNarath1** [comment]
> aw yis i was gonna do exactly this (well some of it). seems you beat me to the draw
> 
> actually i have a couple ideas that i think would be very effective that you missed here. wanna team up or something?

sure thing ideas are good i definitly did overlook stuff so go ahead you can also if you want to work together dm me on discord you should be able to find me on the beestation discord just fine.

File: code/modules/integrated_electronics/core/printer.dm [hunk 1]
```diff
@@ -198,7 +154,75 @@
...
+	return list(
+		get_asset_datum(/datum/asset/spritesheet/circuits),
+	)
+/obj/machinery/integrated_circuit_printer/ui_state(mob/user)
+	return GLOB.physical_state
+
+/obj/machinery/integrated_circuit_printer/ui_interact(mob/user,datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "Circuit_Printer", name)
```
Comment (jlsnow301) [inline code/modules/integrated_electronics/core/printer.dm:167]: Consistency with other tguis
```suggestion
		ui = new(user, src, "CircuitPrinter", name)
```

File: code/modules/integrated_electronics/core/printer.dm [hunk 2]
```diff
@@ -198,7 +154,75 @@
...
+		ui.open()
+
+/obj/machinery/integrated_circuit_printer/ui_data(mob/user)
+	var/list/data = list()
+	data["materials"] = rmat
+	data["program"] = program
+	data["upgrade"] = upgraded
```
Comment (jlsnow301) [inline code/modules/integrated_electronics/core/printer.dm:174]: Ideally we shouldn't lock anything via the client (unless you're just sending this to display "upgraded")
make circuits a list, iterate through the available ones, add an if check for upgraded

File: code/modules/integrated_electronics/core/printer.dm (hunk 1 above)
Comment (PowerfulBacon) [reply code/modules/integrated_electronics/core/printer.dm:167]: This matters on the .js filename

**Comment by PowerfulBacon** [comment]
Im going to be honest since I don't want you to waste a lot of time, but I really do think circuits are a dead cause. Making it so the server stores circuit data fixes one of the many, many circuit issues while introducing many many more.

**Comment by aidan-bowman** [comment]
closed due to inactivity?

**Pull request closed without merging**